### PR TITLE
Gateway - gateway participant update target state in cache

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
https://github.com/apache/helix/issues/2821

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Gateway participant will update cache target state
Improve locking

### Tests

- [X] The following tests are written for this issue:

Update TestHelixGatewayParticipant.java

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-gateway/src/main/java/org/apache/helix/gateway/service/GatewayServiceManager.java [hunk 1]
```diff
@@ -91,20 +91,26 @@ public void onGatewayServiceEvent(GatewayServiceEvent event) {
...
 
+  /**
+   * Overwrite the current state cache with the new current state map, and return the diff of the change.
+   * @param clusterName
+   * @param newCurrentStateMap
+   * @return
+   */
   public  Map<String, Map<String, Map<String, String>>> updateCacheWithNewCurrentStateAndGetDiff(String clusterName,
       Map<String, Map<String, Map<String, String>>> newCurrentStateMap) {
    return  getCache(clusterName).updateCacheWithNewCurrentStateAndGetDiff(newCurrentStateMap);
```
Comment (junkaixue) [inline helix-gateway/src/main/java/org/apache/helix/gateway/service/GatewayServiceManager.java:106]: The naming is not that good. Could be getCacheInstance? What if the get is null? Then would that getOrCreateCacheInstance?

File: helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java [hunk 2]
```diff
@@ -38,7 +42,7 @@ public class GatewayCurrentStateCache {
```
Comment (junkaixue) [inline helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java:42]: Need to be consistent? final on this?

File: helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java [hunk 3]
```diff
@@ -47,111 +51,113 @@ public GatewayCurrentStateCache(String clusterName) {
...
+    }
+    curStateMap.updateWithShardState(resource, shard, shardState);
   }
 
   /**
    * Serialize the target state assignments to a JSON Node.
    * example ： {"instance1":{"resource1":{"shard1":"ONLINE","shard2":"OFFLINE"}}}}
    */
-  public ObjectNode serializeTargetAssignmentsToJSONNode() {
+  public synchronized ObjectNode serializeTargetAssignmentsToJSONNode() {
```
Comment (junkaixue) [inline helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java:119]: Optimize this? Create a snapshot object for get operation? And serailize data and update it when it is needed? I am afraid of having many expired objects increasing GC.

File: helix-gateway/src/main/java/org/apache/helix/gateway/service/GatewayServiceManager.java (hunk 1 above)
Comment (xyuanlu) [reply helix-gateway/src/main/java/org/apache/helix/gateway/service/GatewayServiceManager.java:106]: Yes if get is null, it create instance. I will update name.

File: helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java (hunk 3 above)
Comment (xyuanlu) [reply helix-gateway/src/main/java/org/apache/helix/gateway/util/GatewayCurrentStateCache.java:119]: Then won't create snapshot bringing a lot object for GC?

**Review by zpinto — APPROVED** [review]

**Review by zpinto — APPROVED** [review]
Thanks for changing ObjectNode to be modified instead of recreated :)

**Comment by xyuanlu** [comment]
This change is ready to be merged. Approved by @zpinto 
Commit message: 
Gateway - gateway participant update target state in cache

**Pull request merged**